clouddriver-google/src/main/groovy/com/netflix/spinnaker/clouddriver/google/deploy/handlers/BasicGoogleDeployHandler.groovy

```
@@ -611,46 +629,102 @@ class BasicGoogleDeployHandler implements DeployHandler<BasicGoogleDeployDescrip
+          //
+          // Without waiting for autoscaler creation to complete, subsequent deployment steps (health checks, traffic routing)
+          // may execute before the autoscaler is active, leading to inconsistent behavior and potential deployment failures.
+          //
+          // This fix aligns Spinnaker GCP behavior with Spinnaker AWS behavior, where autoscaling group operations are synchronous.
+          Operation autoscalerOperation = timeExecute(compute.autoscalers().insert(project, zone, autoscaler),
                       "compute.autoscalers.insert",
                       TAG_SCOPE, SCOPE_ZONAL, TAG_ZONE, zone)
+          
+          // Wait for zonal autoscaler creation to complete before proceeding with deployment
+          // Uses GoogleOperationPoller which implements proper retry logic and handles operation status polling
+          googleOperationPoller.waitForZonalOperation(compute, project, zone, autoscalerOperation.getName(),
+            null, task, "autoscaler $serverGroupName", BASE_PHASE)
         }
       }
     }
 
     task.updateStatus BASE_PHASE, "Done creating server group $serverGroupName in $location."
 
-    // Actually update the backend services.
+    // Update backend services and wait for operation completion
     if (willUpdateBackendServices) {
       backendServicesToUpdate.each { BackendService backendService ->
-        safeRetry.doRetry(
+        // Execute backend service update with retry logic for transient errors
+        def operation = safeRetry.doRetry(
           updateBackendServices(compute, project, backendService.name, backendService),
           "Load balancer backend service",
           task,
-          [400, 412],
+          [400, 412],  // Retry on Bad Request (400) and Precondition Failed (412)
           [],
           [action: "update", phase: BASE_PHASE, operation: "updateBackendServices", (TAG_SCOPE): SCOPE_GLOBAL],
           registry
         )
+        
+        if (operation) {
```

Under what circumstances is operation null?  Is it worth logging a message or emitting a metric?

---

The previous invocation would return null always.  Since polling for the state of the operation (i it's there) requires passing oepration information, we have to check if that's null before using the operation.  This null check wasn't needed previously because we weren't polling for state previously.  https://github.com/spinnaker/clouddriver/blob/aa0c51ae74351bd3ae0ce9949212bb5ac5ae073c/clouddriver-google/src/main/java/com/netflix/spinnaker/clouddriver/google/deploy/SafeRetry.java#L89 and below where null was always returned.

---

I see....Wouldn't the world be a better place if SafeRetry.doRetry let the exception bubble up?  Seems like we're effectively swallowing them which doesn't seem great.

---

Yes I agree ;) I think that's another area that could use some improvement ... probably separate from this.